python/pyarrow/scalar.pxi

```diff
@@ -1027,6 +1027,53 @@ cdef class ExtensionScalar(Scalar):
         return pyarrow_wrap_scalar(<shared_ptr[CScalar]> sp_scalar)
 
 
+cdef class VariableShapeTensorScalar(ExtensionScalar):
+    """
+    Concrete class for variable shape tensor extension scalar.
+    """
+
+    def to_numpy_ndarray(self):
+        # TODO: allow any permutation
+        """
+        Convert variable shape tensor extension scalar to a numpy array.
+        Note: ``permutation`` should be trivial (``None`` or ``[0, 1, ..., len(shape)-1]``).
+        """
+
+        if self.type.permutation is None or self.type.permutation == list(range(len(self.type.permutation))):
+            shape = self.value[0].values.to_pylist()
+            np_flat = np.asarray(self.value[1].values)
+            return np_flat.reshape(shape)
+        else:
+            raise ValueError(
+                'Only non-permuted tensors can be converted to numpy tensors.')
+
+    def to_tensor(self):
+        """
+        Convert variable shape tensor extension scalar to a pyarrow.Tensor.
+        """
+        cdef:
+            shared_ptr[CTensor] ctensor
+            vector[int64_t] strides
+            vector[c_string] dim_names
+
+            shared_ptr[CVariableShapeTensorType] typ = static_pointer_cast[CVariableShapeTensorType, CDataType](
+                self.wrapped.get().type)
+
+            shared_ptr[CDataType] ty = typ.get().value_type()
+            # TODO: this accesses the full buffer instead of a slice
+            shared_ptr[CBuffer] data = pyarrow_unwrap_buffer(self.value[1].values.buffers()[1])
```

Comment:
@jorisvandenbossche is it expected behavior that this returns the full array underlying buffer instead of the scalar underlying slice? Is there a way to get access to the slice or the offset here? Currently this always returns the first scalar in the array which is not terribly useful.
cc @AlenkaF

Reply:
cc @pitrou

Reply:
Why are you not doing this in C++ instead? This would probably allow doing what you want and would also be useful to non-Python libraries.

Reply:
This does add a `GetTensor(i)` method, but it can't be called "from within" Python's `ExtensionScalar` as far as I know.

Reply:
Why not define a `GetTensor` operating on a scalar?
```c++
class ARROW_EXPORT VariableShapeTensorType : public ExtensionType {
 public:
   Result<std::shared_ptr<Tensor>> GetTensor(const ExtensionScalar&);
```

Reply:
But how would it be applied in Python? Getting a scalar in Python just returns an `ExtensionScalar` that is unaware of it's offset and hence can't correctly slice from array's underlying buffer. Maybe I'm missing something :D

Reply:
The `ExtensionScalar` has an underlying storage scalar, no?

Reply:
If I do this: `shared_ptr[CBuffer] data = pyarrow_unwrap_buffer(self.value[1].values.buffers()[1])` I get the whole buffer of the underlying array instead of a slice that I'd expect to be the scalar. Is there a way to gust get the buffer slice?

Reply:
I think we're talking past each other here.

Reply:
Sorry :)

I've moved the python `get_tensor` logic to C++ (https://github.com/apache/arrow/pull/38008/commits/7e054f7a29243d372fa28f7430f6e31d191a6162).
Here's the `VariableShapeTensorType.GetTensor(ExtensionScalar)` method. Unfortunately it also always gets values from the beginning of the buffer instead offset of the Scalar. I imagine there's a way to somehow grab the correct buffer slice that I'm missing. See line 220.

https://github.com/apache/arrow/blob/7e054f7a29243d372fa28f7430f6e31d191a6162/cpp/src/arrow/extension/variable_shape_tensor.cc#L198-L224

Reply:
[Example error](https://github.com/apache/arrow/actions/runs/7013790354/job/19080485614?pr=38008#step:6:6320). Any suggestions how to access the correct slice?

Reply:
You simply have to apply the array offset to the buffer pointer, unless I'm missing something. If you're ignoring the offset, you'll certainly get the wrong data area indeed.

Reply:
Done!
I was surprised that we default to full buffer there or at least that slice is not easily available.

Reply:
Do you think it would make sense to provide a `Scalar.GetSlicedBuffer()`?